### PR TITLE
Adds Announcements to WebSDK

### PR DESCRIPTION
**Pull request by michaelkantor**

### CODE REVIEW

File: src/syncable.js [hunk 1]
```diff
@@ -144,17 +205,30 @@ Syncable.prototype.syncState = Constants.SYNC_STATE.NEW;
 Syncable.prototype._syncCounter = 0;
 
 /**
+ * Prefix to use when triggering events
+ */
+Syncable.eventPrefix = '';
```
Comment (michaelkantor) [inline src/syncable.js]: Lets Syncable trigger events with the right name; have filed a jira to use this more widely.

File: src/syncable.js [hunk 2]
```diff
@@ -144,17 +205,30 @@ Syncable.prototype.syncState = Constants.SYNC_STATE.NEW;
...
   },
 });
 
+/**
+ * Array of classes that are subclasses of Syncable.
+ *
+ * Used by Factory function.
+ * @private
+ */
+Syncable.subclasses = [];
```
Comment (michaelkantor) [inline src/syncable.js]: Allows Syncable to act as a Factory for any subclasses that register.

File: src/websockets/change-manager.js [hunk 3]
```diff
@@ -118,10 +118,17 @@ class WebsocketChangeManager {
...
+    } else {
+      switch (Utils.typeFromID(msg.object.id)) {
+        case 'conversations':
+          if (Conversation._loadResourceForPatch(msg.data)) this.client.getConversation(msg.object.id, true);
+          break;
+        case 'messages':
+          if (Message._loadResourceForPatch(msg.data)) this.client.getMessage(msg.object.id, true);
+          break;
+        case 'announcements':
+          break;
```
Comment (michaelkantor) [inline src/websockets/change-manager.js]: A change event on an Announcement we haven't loaded? Ignore it.

File: src/root.js [hunk 4]
```diff
@@ -218,7 +218,9 @@ class Root extends EventClass {
 
     // Iterate over all formally defined properties
     try {
-      const keys = Object.keys(this.constructor.prototype);
+      const keys = [];
+      for (let key in this.constructor.prototype) if (!(key in layer.Root.prototype)) keys.push(key);
```
Comment (michaelkantor) [inline src/root.js]: Object.keys() fails to get parent class properties, and is really only effective for looking at POJOs

File: src/message.js [hunk 5]
```diff
@@ -920,86 +926,23 @@ class Message extends Syncable {
...
-   * This will get the Message from cache or layer.Message.load it from the server if not cached.
-   * Typically you do not need to call this method directly.
-   *
-   * @method load
-   * @static
-   * @param  {string} id - Message identifier
-   * @param  {layer.Client} client - Client whose conversations should contain the new message
-   * @return {layer.Message}
-   */
-  static load(id, client) {
```
Comment (michaelkantor) [inline src/message.js]: Moved to Syncable

File: src/message.js [hunk 6]
```diff
@@ -920,86 +926,23 @@ class Message extends Syncable {
...
-    if (!client || !(client instanceof Root)) throw new Error(LayerError.dictionary.clientMissing);
-    if (id.indexOf('layer:///messages/') !== 0) throw new Error(LayerError.dictionary.invalidId);
-
-    const message = new Message({
-      id,
-      url: client.url + id.substring(8),
+  static _createFromServer(message, client) {
```
Comment (michaelkantor) [inline src/message.js]: Now returns Message instead of {message: message, new: isNew}

File: src/message.js [hunk 7]
```diff
@@ -174,6 +173,10 @@ class Message extends Syncable {
     this.isInitializing = false;
     if (options && options.fromServer) {
       client._addMessage(this);
+      const status = this.recipientStatus[client.userId];
+      if (status !== Constants.RECEIPT_STATE.READ && status !== Constants.RECEIPT_STATE.DELIVERED) {
+        this._sendReceipt('delivery');
```
Comment (michaelkantor) [inline src/message.js]: Was in createFromServer; don't recall why this moved here.

File: src/conversation.js [hunk 8]
```diff
@@ -1550,5 +1472,5 @@ Conversation._supportedEvents = [
   'conversations:change'].concat(Syncable._supportedEvents);
 
 Root.initClass.apply(Conversation, [Conversation, 'Conversation']);
-
+Syncable.subclasses.push(Conversation);
```
Comment (michaelkantor) [inline src/conversation.js]: Registers for use by Factory

File: src/conversation.js [hunk 9]
```diff
@@ -914,41 +914,8 @@ class Conversation extends Syncable {
...
-   *
-   * Called from the static layer.Conversation.load() method
-   *
-   * @method _load
-   * @private
-   */
-  _load() {
-    this.syncState = Constants.SYNC_STATE.LOADING;
-    this._xhr({
-      url: '',
```
Comment (michaelkantor) [inline src/conversation.js]: Moved to Syncable

File: src/syncable.js [hunk 10]
```diff
@@ -19,10 +19,71 @@
  */
 
 const Root = require('./root');
-const Constants = require('./const');
+const { SYNC_STATE } = require('./const');
+const LayerError = require('./layer-error');
 
 class Syncable extends Root {
 
+  static load(id, client) {
```
Comment (michaelkantor) [inline src/syncable.js]: Identity branch starts to explore checking the database BEFORE loading from the server.

File: src/query.js [hunk 11]
```diff
@@ -482,6 +485,44 @@ class Query extends Root {
...
+    }
+    this.isReset = false;
+
+    // This is a pagination rather than an initial request if there is already data; get the fromId
+    // which is the id of the last result.
+    const lastMessage = this.data[this.data.length - 1];
+    const lastMessageInstance = !lastMessage ? null : this._getInstance(lastMessage);
+    const fromId = (lastMessageInstance && lastMessageInstance.isSaved() ? '&from_id=' + lastMessageInstance.id : '');
+
+    // If the last message we have loaded is already the Conversation's lastMessage, then just request data without paging,
```
Comment (nilgradisnik) [inline src/query.js]: This comment doesn't make any sense here.

File: src/query.js [hunk 12]
```diff
@@ -940,9 +988,17 @@ class Query extends Root {
...
     const list = evt.messages
-                  .filter(message => message.conversationId === this._predicate)
-                  .filter(message => this._getIndex(message.id) === -1)
-                  .map(message => this._getData(message));
+      .filter(message => {
+        const type = Util.typeFromID(message.id);
+        return type === 'messages' && this.model === MESSAGE ||
+                type === 'announcements' && this.model === ANNOUNCEMENT;
+      })
+      .filter(message => {
```
Comment (nilgradisnik) [inline src/query.js]: Why are you filtering twice here? It looks like you can incorporate this `type === 'announcements' || message.conversationId === this._predicate` into the first filter method.

File: src/syncable.js [hunk 13]
```diff
@@ -19,10 +19,71 @@
...
+
+    const ConstructorClass = Syncable.subclasses.filter(aClass => obj.id.indexOf(aClass.prefixUUID) === 0)[0];
+    const syncItem = new ConstructorClass(obj);
+
+    syncItem._load();
+    return syncItem;
+  }
+
+  /**
+   * Load this conversation from the server.
```
Comment (nilgradisnik) [inline src/syncable.js]: This is not specific to conversation anymore. Let's updated these comments to something more generic.

**Pull request merged**